### PR TITLE
[RFC] ASoC: SOF: topology: use set_get_data in process load

### PR DESCRIPTION
**Pull request by juimonen**
Currently when loading sof process components there's a check if binary
control data is associated with it. If found the data is extracted to be
part of component loading and initialization. If binary data exceeds the
ipc max size, loading fails with error as large message support is only
implemented in set_get_data method. So make the process loading use
set_get_data to enable large parameters in component initialization.

Signed-off-by: Jaska Uimonen <jaska.uimonen@intel.com>

### CODE REVIEW

**Comment by juimonen** [comment]
I'm not sure if there's restrictions why we could not send params during component loading...
corresponding FW PR:
https://github.com/thesofproject/sof/pull/1408

File: include/uapi/sound/sof/abi.h [hunk 1]
```diff
@@ -26,7 +26,7 @@
 
 /* SOF ABI version major, minor and patch numbers */
 #define SOF_ABI_MAJOR 3
-#define SOF_ABI_MINOR 5
+#define SOF_ABI_MINOR 6
```
Comment (plbossart) [inline include/uapi/sound/sof/abi.h]: is this needed?

File: sound/soc/sof/topology.c [hunk 2]
```diff
@@ -1738,63 +1738,18 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
 	struct sof_abi_hdr *pdata = NULL;
 	struct sof_ipc_comp_process *process;
 	size_t ipc_size, ipc_data_size = 0;
-	int ret, i, offset = 0;
+	int ret, i;
+	int ipc_cmd, ctrl_type;
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: reverse xmas tree and one declaration per line if possible.

File: sound/soc/sof/topology.c [hunk 3]
```diff
@@ -1738,63 +1738,18 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
...
-		if (pdata->magic != SOF_ABI_MAGIC)
-			continue;
-
-		ipc_data_size += pdata->size;
-	}
-
 	ipc_size = sizeof(struct sof_ipc_comp_process) +
 		le32_to_cpu(private->size) +
-		ipc_data_size;
+	  ipc_data_size;
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: alignment?

Comment (plbossart) [inline sound/soc/sof/topology.c]: I don't get at all what this simplification does, we now have ipc_data_size = 0?

File: sound/soc/sof/topology.c [hunk 4]
```diff
@@ -1819,6 +1774,17 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
 
 	sof_dbg_comp_config(scomp, &process->config);
 
+	process->size = ipc_data_size;
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: ipc_data_size is still zero here?

**Comment by singalsu** [comment]
@cujomalainey This work fixes the embedding of long EQ filter coefficients set into topology file.

**Comment by cujomalainey** [comment]
Much appreciated, once we have this on a chromebook I will finish up testing on the GUI then send a PR.

File: sound/soc/sof/topology.c [hunk 5]
```diff
@@ -1819,6 +1774,17 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
...
+				 ipc_size, r, sizeof(*r));
+
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: failed to create process comp\n");
+		goto err;
+	}
+
 	/*
 	 * found private data in control, so copy it.
 	 * get possible component controls - get size of all pdata,
```
Comment (ranj063) [inline sound/soc/sof/topology.c:1902]: @juimonen the comment is no longer valid.

File: sound/soc/sof/topology.c [hunk 6]
```diff
@@ -1856,17 +1824,39 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
...
 
-		memcpy(&process->data + offset, pdata->data, pdata->size);
-		offset += pdata->size;
+		switch (scontrol->cmd) {
+		case SOF_CTRL_CMD_VOLUME:
+		case SOF_CTRL_CMD_ENUM:
+		case SOF_CTRL_CMD_SWITCH:
+			ipc_cmd = SOF_IPC_COMP_SET_VALUE;
+			ctrl_type = SOF_CTRL_TYPE_VALUE_CHAN_SET;
+			break;
```
Comment (ranj063) [inline sound/soc/sof/topology.c]: @juimonen I'm not sure about this. We cant pass any private data in the kcontrol as a volume set command. Either there has to be a sanity check to make sure it is in the volume table or we remove it. I think it is unnecessary to pass any other data that the binary control data.

**Comment by ranj063** [comment]
@juimonen Did you test this change with the kwd topology. The selector and the KPB components expect binary data to be attached to the process ipc struct payload. This change will break that requirement

**Comment by juimonen** [comment]
@ranj063 I did not test it with KWD. But KWD should also support large data and message sending? I mean the whole large data was done for KWD...?

File: sound/soc/sof/topology.c (hunk 6 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: ok

File: sound/soc/sof/topology.c (hunk 5 above)
Comment (juimonen) [reply sound/soc/sof/topology.c:1902]: ok

**Comment by juimonen** [comment]
@plbossart @ranj063 thanks for reviewing. it is me leaving friday evening and quickly pushing something for others to see... the abi change is pointless, I didn't notice it, I only need it for my local testing. ipc_data_size should be also removed. 

I only wanted to get first high level comments is there some conceptual problem of not using set_get_data in component creation. Without fw side PR this change will actually oops the DSP...

So fw/dsp implicitly assumes that fw side component should have valid pipeline pointer when these messages are sent. Neither side is checking it. if I understand correctly the pipeline pointer is valid only when pipeline is "complete".  But I don't get why we should restrict of sending controls only when pipeline is ready...

**Comment by juimonen** [comment]
@plbossart with this PR I want to enable large messages in component creation.

**Comment by ranj063** [comment]
@juimonen we talked about sending default configuration parameters along as binary data as ipc payload when new components are created. So for components like selector or the kpb, we should think about adding the config data to the widget instead if having a binary control.

File: include/uapi/sound/sof/abi.h (hunk 1 above)
Comment (juimonen) [reply include/uapi/sound/sof/abi.h]: no

File: sound/soc/sof/topology.c (hunk 2 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: ok

File: sound/soc/sof/topology.c (hunk 3 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: ok

Comment (juimonen) [reply sound/soc/sof/topology.c]: removed totally.

File: sound/soc/sof/topology.c (hunk 4 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: removed.

**Comment by juimonen** [comment]
@ranj063 my point here is: if you use controls to have binary data, you have to be prepare in component that someone is actually using the control, so not just in creation phase...?

File: sound/soc/sof/topology.c [hunk 7]
```diff
@@ -1812,61 +1765,83 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
...
-	 * found private data in control, so copy it.
-	 * get possible component controls - get size of all pdata,
-	 * then memcpy with headers
+	 * Get possible binary controls, which might have private data.
+	 * Then parse the size of the data and send it to dsp.
 	 */
 	for (i = 0; i < widget->num_kcontrols; i++) {
 		kc = &widget->kcontrol_news[i];
 
+		/* we only care about bytes control */
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: why? Isn't that a restriction compared to the previous solution?

File: sound/soc/sof/topology.c [hunk 8]
```diff
@@ -1728,73 +1728,26 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
...
-		if (pdata->magic != SOF_ABI_MAGIC)
-			continue;
-
-		ipc_data_size += pdata->size;
-	}
-
 	ipc_size = sizeof(struct sof_ipc_comp_process) +
-		le32_to_cpu(private->size) +
-		ipc_data_size;
+		le32_to_cpu(private->size);
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: that one I just can't figure out? Looks to me like things may work for binary controls but will no longer work for other types of controls.
Confused.

File: sound/soc/sof/topology.c (hunk 7 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: @plbossart currently we are using component related control's binary blob to convey binary parameters to "process" component. and yeah, the name was not my idea...

As @ranj063 pointed out in another comment, why to even try to send other control types like enums or volume data at this point because it doens't make sense? Hence, I only try to find binary and "continue" with other types, if someone has tried to push that from topology, for some strange reason.

File: sound/soc/sof/topology.c (hunk 8 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: @plbossart see above. but I'm not totally sure about this, it is very easy to put back the parsing for other types of data. @ranj063 what do you think?

**Comment by lyakh** [comment]
I'm probably more confused by this than anybody else, trying to review this. Isn't this commit trying to do several things? Changing the logic and increasing the maximum size of data, that can be processed? Shouldn't at least they be separated?

**Comment by juimonen** [comment]
@lyakh so previously data was dig out from component related control and copied to component ipc struct. Now as large messages is only supported with set_get_data, this previous method can't send big parameters as part of the component creation struct. This PR is just removing this copy, so component is created with empty binary data and after it is created I'm sending the private binary data with set_get_data (which support large messages). Instead of previously only sending 1 ipc in creation, I'm sending 2 ipc (component creation then the data). 

I think this is quite simple. The only thing to "understand" that we might have component related private data in its related control. This was originally done because it is really nasty to include both tokens and binary blob to component in topology, so it is nasty from kernel point of view when parsing this. There needed to be all kinds of hard coded offsets etc. So we decided to add the possible binary blob to related control. Despite this, if you remember, I was specifically told to implement the large messages only to control set/get data, so not universally (my first PR was for universal IPC and it was not accepted) -> if we try to create component bigger than max ipc size it will fail -> let's use the set_get_data in creation so we can send bigger params.

In the end, the most "correct" way of doing this would be to 1) enable adding binary blob together with tokens to component topology and parsing it nicely. Then 2) adding large message support for component creation. But this is not as easy and straightforward as this PR...

File: sound/soc/sof/topology.c (hunk 8 above)
Comment (plbossart) [reply sound/soc/sof/topology.c]: I don't see what's different between the various types of kcontrols. If you can set them in topology then it should be the same handling for all.

Comment (juimonen) [reply sound/soc/sof/topology.c]: @plbossart, ok fine by me, i'll add them back

**Comment by keyonjie** [comment]
This looks legit to me, though it will break today's KWD, as we don't have set_get_data support for kpb and selector ATM.
@mrajwa @lbetlej we might need add set_get_data support there(similar to detector component).

**Comment by juimonen** [comment]
@keyonjie ok, I see detector and selector support this but not kpb..

**Comment by juimonen** [comment]
@plbossart @ranj063 @keyonjie updated. This version supports legacy components, so it is backwards compatible. Data fitting to single ipc is sent with "old method". There's also a 2nd commit which refactors the process loading function. It it looks fine it can be squashed to the first one. Just to make review easier...

File: sound/soc/sof/topology.c [hunk 9]
```diff
@@ -1794,11 +1799,17 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
...
+		ipc_size -= ipc_data_size;
+		ipc_data_size = 0;
+	}
+
 	process = kzalloc(ipc_size, GFP_KERNEL);
 	if (!process)
 		return -ENOMEM;
 
-	/* configure iir IPC message */
+	/* configure process IPC message */
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: nit-pick: should be separate patch?

File: sound/soc/sof/topology.c [hunk 10]
```diff
@@ -1858,13 +1873,77 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
...
-	if (ret >= 0)
+
+	if (ret >= 0 && ipc_data_size)
 		return ret;
+
+	/* send control data with large message supported method */
+	for (i = 0; i < widget->num_kcontrols; i++) {
+		kc = &widget->kcontrol_news[i];
+
+		/* we only care about bytes control */
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: you should elaborate on this, the comment is just too short to understand what you explained in the previous review (it doesn't make sense to send the other types through topology).

File: sound/soc/sof/topology.c [hunk 11]
```diff
@@ -1858,13 +1873,77 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
...
+				widget->dobj.widget.kcontrol_type);
+			ret = -EINVAL;
+			goto err;
+		}
+
+		/* don't include if there's no private data */
+		if (!scontrol->control_data->data)
+			continue;
+
+		/* make sure data is valid - data can be updated at runtime */
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: is the comment correct? I thought we are in the part where we send stuff from topology, so what can be updated?

File: sound/soc/sof/topology.c [hunk 12]
```diff
@@ -1821,59 +1858,30 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
 			       ARRAY_SIZE(comp_tokens), private->array,
 			       le32_to_cpu(private->size));
 	if (ret != 0) {
-		dev_err(sdev->dev, "error: parse process.cfg tokens failed %d\n",
+		dev_err(sdev->dev, "error: parse process tokens failed %d\n",
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: separate patch?

File: sound/soc/sof/topology.c [hunk 13]
```diff
@@ -1749,50 +1825,11 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
 		return -EINVAL;
 	}
 
-	/*
-	 * get possible component controls - get size of all pdata,
-	 * then memcpy with headers
-	 */
+	/* get possible component controls and get size of all pdata */
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: separate patch? is this change needed?

File: sound/soc/sof/topology.c [hunk 14]
```diff
@@ -1719,6 +1719,85 @@ static int sof_widget_load_siggen(struct snd_soc_component *scomp, int index,
 	return ret;
 }
 
+static int sof_get_control_data(struct snd_sof_dev *sdev,
```
Comment (plbossart) [inline sound/soc/sof/topology.c]: @juimonen general comment on this comment:
On one hand I appreciate the logic of trying to make the change then refactor
On the other hand, this looks like tick-tock patches where you change what you just introduced.
Maybe make sure that this a block change and avoid additional changes such as changed comments or changed sequences, etc?

**Comment by ranj063** [comment]
@juimonen sorry for the delay. I like the refactoring part but I have a suggestion to make it a bit more intuitive. We still call the sof_get_control_data() method 3 times. We should only call it once and gather all the information in one go. Basically, it should return the total size of the data for all controls as the return value. Additionally, it should create a pdata and scontrol array of size num_controls. We could possibly create a struct like this (for example):

struct widget_data {
void **pdata;
struct snd_sof_control **control;
int ipc_cmd;
int cmd_type;
};

and pass it to the sof_get_control_data() and populate everything we need. Once this is done, you can determine the course of action based on the returned size.

**Comment by juimonen** [comment]
@ranj063 heh yes, was thinking on the same lines, just little bit worried of using dynamic arrays in kernel code (don't have so much experience on this) and then thinking is it a overkill for removing 3 iterations? But agree that would be the cleanest solution...

**Comment by ranj063** [comment]
> then thinking is it a overkill for removing 3 iterations? But agree that would be the cleanest solution...

@juimonen I guess the problem right now is that no one other than you and me understand why this is being done. And calling the same function thrice is only going to confuse people more. Whereas if we do it once, we can add proper comments to say that we're calculating the total size, copying the data and saving the control data for IPC's, it will make things clearer.

**Comment by juimonen** [comment]
@ranj063 yes sure, i'll try to do it like that.

File: sound/soc/sof/topology.c (hunk 12 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: well, yeah I'm just fixing things what check-patch wants me to do, I should then do intermediate commits without hooks. These are old stuff that checkpatch doesn't like. Anyway I can put all comment changes to separate patch.

File: sound/soc/sof/topology.c (hunk 9 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: ok

File: sound/soc/sof/topology.c (hunk 11 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: this is an old comment so this is mystery to me also... so I know that data can be updated runtime but it doesn't too much connect to "validity". I'll fix it in separate commit.

File: sound/soc/sof/topology.c (hunk 10 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: actually this is now just wrong, will remove.

**Comment by juimonen** [comment]
@plbossart @ranj063 updated to Ranjani's comments. I tried to leave any extra touching of old comments away...not sure it totally succeeded. As said I think these patches should be anyway squashed...

File: sound/soc/sof/topology.c [hunk 15]
```diff
@@ -1858,13 +1873,76 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
...
 
+process:
 	process->size = ipc_data_size;
 	swidget->private = process;
 
 	ret = sof_ipc_tx_message(sdev->ipc, process->comp.hdr.cmd, process,
 				 ipc_size, r, sizeof(*r));
-	if (ret >= 0)
+
+	if (ret >= 0 && ipc_data_size)
```
Comment (ranj063) [inline sound/soc/sof/topology.c]: @juimonen we should also return is ret < 0 because that means the ipc message has failed. so I think a check for if (ipc_data_size) should suffice here.

File: sound/soc/sof/topology.c [hunk 16]
```diff
@@ -1777,22 +1760,81 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
...
+				GFP_KERNEL);
+
+		if (!wdata)
+			return -ENOMEM;
+
+		/* get possible component controls and get size of all pdata */
+		ipc_data_size = sof_get_control_data(sdev, widget, wdata);
 
-		ipc_data_size += pdata->size;
+		if (ipc_data_size <= 0) {
```
Comment (ranj063) [inline sound/soc/sof/topology.c:1857]: @juimonen is it also an error is ipc_data_size is 0? if the process comp has no controls, this will be 0 isnt it?

File: sound/soc/sof/topology.c [hunk 17]
```diff
@@ -1828,122 +1872,45 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
...
+		ret = snd_sof_ipc_set_get_comp_data(sdev->ipc, wdata[i].control,
+						    wdata[i].ipc_cmd,
+						    wdata[i].ctrl_type,
+						    wdata[i].control->cmd,
+						    true);
 		if (ret != 0) {
 			dev_err(sdev->dev, "error: send control failed\n");
 			goto err;
 		}
 	}
```
Comment (ranj063) [inline sound/soc/sof/topology.c:1947]: @juimonen there's an error in freeing process here. kfree(process) would be called even if there's no error which will results in swidget->private being NULL and this will cause the component set up fail upon resume.

File: sound/soc/sof/topology.c (hunk 15 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: @ranj063 hmm ok this is dodgy... I'll refine it and remove the goto's to make it clear.

File: sound/soc/sof/topology.c (hunk 16 above)
Comment (juimonen) [reply sound/soc/sof/topology.c:1857]: @ranj063 I think we should not be here if there's no controls... ? this is under:
if (widget->num_kcontrols) {

File: sound/soc/sof/topology.c (hunk 17 above)
Comment (juimonen) [reply sound/soc/sof/topology.c:1947]: @ranj063 yes good catch, I'll remove the goto's because they make this ugly. I could do this with couple of goto's but that start's to be too much

**Comment by plbossart** [comment]
@juimonen can you address comments and mark them as resolved when applicable. It's difficult to review when we don't know if the comments are stale or still apply. Thanks!

File: sound/soc/sof/topology.c (hunk 14 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: @plbossart I tried to remove my changes to all non-related comments even they would be originally erroneous. That can be addressed in another pr.

File: sound/soc/sof/topology.c (hunk 15 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: ok

File: sound/soc/sof/topology.c (hunk 13 above)
Comment (juimonen) [reply sound/soc/sof/topology.c]: ok, removed the change

**Comment by juimonen** [comment]
@ranj063 you have time still to take a look? thanks!

**Comment by ranj063** [comment]
@juimonen looks good now but I think these should be squashed or the refactor done first.

**Comment by juimonen** [comment]
@ranj063 yes I made the separate re-factor for @plbossart (cause I know he likes them separately). @plbossart if you're good, I'll squash these commits together?

**Comment by plbossart** [comment]
I can't look into this in details so what @ranj063 says

File: sound/soc/sof/topology.c [hunk 18]
```diff
@@ -1822,51 +1879,49 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
...
+		kfree(wdata);
+		kfree(process);
 		return ret;
-err:
-	kfree(process);
+	}
+
+	/* we sent the data in single message so return */
+	if (ipc_data_size) {
+		kfree(wdata);
```
Comment (plbossart) [inline sound/soc/sof/topology.c:1929]: is there a missing kfree(process) here? 
Or is this intentional, and might be worth documenting.

File: sound/soc/sof/topology.c [hunk 19]
```diff
@@ -1822,51 +1879,49 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
...
+	for (i = 0; i < widget->num_kcontrols; i++) {
+		wdata[i].control->readback_offset = 0;
+		ret = snd_sof_ipc_set_get_comp_data(sdev->ipc, wdata[i].control,
+						    wdata[i].ipc_cmd,
+						    wdata[i].ctrl_type,
+						    wdata[i].control->cmd,
+						    true);
+		if (ret != 0) {
+			dev_err(sdev->dev, "error: send control failed\n");
+			kfree(process);
```
Comment (plbossart) [inline sound/soc/sof/topology.c:1943]: and this looks wrong, you might free the same thing multiple times in the loop - that may be harmless but it is a telltale sign of memory allocation issues. if there a missing return?

File: sound/soc/sof/topology.c [hunk 20]
```diff
@@ -1822,51 +1879,49 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
...
+		}
+	}
+
+	kfree(wdata);
```
Comment (plbossart) [inline sound/soc/sof/topology.c:1948]: and here process is not freed either

File: sound/soc/sof/topology.c (hunk 18 above)
Comment (ranj063) [reply sound/soc/sof/topology.c:1929]: @plbossart this is intentional. process is saved as swidget->private for loading the comp upon resuming from PM suspend

File: sound/soc/sof/topology.c (hunk 19 above)
Comment (ranj063) [reply sound/soc/sof/topology.c:1943]: @plbossart good catch. we need a break after kfree(process). @juimonen is out for the next week or so. I will update the PR on his behalf.

**Comment by juimonen** [comment]
@plbossart so "wdata" is local dynamic variable which should be freed after errors and end of function. I have no idea how many controls this process has and I don't want to put there static array. 

"process" is not local so it should be freed only after error. 

As said I changed this to get rid of ugly goto's and the missing break in the loop you found is a real issue and fixed now by @ranj063

File: sound/soc/sof/topology.c (hunk 19 above)
Comment (juimonen) [reply sound/soc/sof/topology.c:1943]: ok

**Comment by juimonen** [comment]
@ranj063 @plbossart updated the break also here.

**Comment by juimonen** [comment]
SOFCI TEST

**Comment by ranj063** [comment]
@juimonen could you please rebase ?

**Comment by juimonen** [comment]
@ranj063 rebased

**Comment by ranj063** [comment]
@juimonen still shows "This branch is out-of-date with the base branch"?

**Comment by juimonen** [comment]
@ranj063 oh accidentally pushed the branch to sof repo than my own. should be now rebased, will delete the branch from origin...

**Comment by ranj063** [comment]
@juimonen unfortunately @plbossart has to approve this PR before it can be merged because he had requested changes.

**Comment by juimonen** [comment]
@ranj063 no worries and thanks for making the extra PR effort.

**Comment by juimonen** [comment]
@ranj063 @plbossart rebased. I've no idea why some pm_status check is failing...

File: sound/soc/sof/topology.c [hunk 21]
```diff
@@ -1795,31 +1783,98 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
...
+	}
+
 	process = kzalloc(ipc_size, GFP_KERNEL);
-	if (!process)
+	if (!process) {
+		kfree(wdata);
 		return -ENOMEM;
+	}
 
 	/* configure iir IPC message */
```
Comment (plbossart) [inline sound/soc/sof/topology.c:1879]: could we please fixup this comment in a follow-up patch? Thanks!
